### PR TITLE
fix for pcolormesh doesn't allow shading = 'flat' in the option

### PR DESCRIPTION
**Pull request by vivekvedant**
Fix for pcolormesh doesn't allow shading = 'flat' in the option #24678

## PR Checklist

**Documentation and Tests**
- [x] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by melissawm** [comment]
Hi @vivekvedant - thanks for the PR! It looks like you have a couple of linting errors (you can see them here https://github.com/matplotlib/matplotlib/actions/runs/3980860096/jobs/6824208412 or clicking on "Details" on the Linting/flake8 job.) This does not block a review of your PR but if you have the time it would be nice to fix them 😄 Otherwise, now we wait for a maintainer to review your code. Thanks!

File: lib/matplotlib/axes/_axes.py
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py:5752]: This error should probably say something about shading='flat', because it is possible to pass x and y the same shape as C for shading='auto' and shading='nearest'.  See https://matplotlib.org/stable/gallery/images_contours_and_fields/pcolormesh_grids.html for a discussion.

Comment (vivekvedant) [reply lib/matplotlib/axes/_axes.py]: @jklymak Sir, Updated the error message

File: lib/matplotlib/axes/_axes.py
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py:5751]: Y should be capitalized.  Suggest "Relevant" because X and Y can be 2-D....  Otherwise this looks good to me...
```suggestion
                raise TypeError(f"Relevant dimensions of X({Nx}) and Y({Ny}) should"
```

**Review by jklymak — APPROVED** [review]

**Comment by ksunden** [comment]
lib/matplotlib/tests/test_axes.py::test_pcolorflaterror

Test failure is a consequence of this change, specifically that the older error message pattern is being matched in the test.

File: lib/matplotlib/axes/_axes.py
Comment (vivekvedant) [reply lib/matplotlib/axes/_axes.py:5751]: Ok Sir, Thank you for the suggestion

**Comment by vivekvedant** [comment]
Ok @ksunden Sir, Thanks for the insights I will recommit the changes to solve this issue

**Comment by vivekvedant** [comment]
@jklymak Sir, I have updated the error message and also fix the issue related to Test failure

**Review by dstansby — COMMENTED** [review]
Thanks a lot for the PR - looks great to me! 👍

**Review by dstansby — APPROVED** [review]

**Comment by jklymak** [comment]
Thanks @vivekvedant and congratulations on your first Matplotlib PR.   We hope to see you again...

**Pull request merged**
